{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "53c35b01-463b-4634-926f-b54ddcad1f6e",
   "metadata": {},
   "source": [
    "# Workflow for processing sequencing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43b504d1-6a92-4e2c-9bec-4214cdf6dc6b",
   "metadata": {},
   "source": [
    "## Quality control"
   ]
  },
  {
   "cell_type": "raw",
   "id": "873adade-64ef-45ef-a30e-1b2458da62d4",
   "metadata": {},
   "source": [
    "00_pretrim_fastqc.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a9f80c81-289c-44f7-8965-0eaed2324dba",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=pretrim_fastqc\n",
    "#SBATCH --array=01-25\n",
    "#SBATCH --time=05:00:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --output=pretrim_fastqc_%A_%a.out\n",
    "#SBATCH --error=pretrim_fastqc_%A_%a.err\n",
    "\n",
    "#load modules\n",
    "module load FastQC/0.12.1\n",
    "module load MultiQC\n",
    "\n",
    "# ppaths\n",
    "RAW_FASTQ_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/raw_fastqc_results\"\n",
    "\n",
    "mkdir -p $OUTPUT_DIR\n",
    "\n",
    "# run fastqc\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# Run FastQC on R1 and R2 files\n",
    "for lane in {1..4}; do\n",
    "    fastqc $RAW_FASTQ_DIR/AC${SAMPLE_ID}_S*_L00${lane}_R1_001.fastq.gz \\\n",
    "\t$RAW_FASTQ_DIR/AC${SAMPLE_ID}_S*_L00${lane}_R2_001.fastq.gz \\\n",
    "        --outdir=$OUTPUT_DIR\n",
    "\n",
    "    echo \"FastQC analysis completed for sample ${SAMPLE_ID} in lane ${lane}\"\n",
    "done\n",
    "\n",
    "echo \"FastQC reports have been generated in $OUTPUT_DIR\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "747ccdcf-203b-4685-9f42-490aba047002",
   "metadata": {},
   "source": [
    "00_raw_reads_multiqc.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "f64dca52-3b56-4211-92fc-d1b3c7083e9b",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=raw_reads_multiqc\n",
    "#SBATCH --time=03:00:00\n",
    "#SBATCH --mem=6GB\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --output=raw_reads_multiqc_%A.out\n",
    "#SBATCH --error=raw_reads_multiqc_%A.err\n",
    "\n",
    "# load modules\n",
    "module load MultiQC\n",
    "\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/raw_fastqc_results\"\n",
    "\n",
    "# run the programme\n",
    "multiqc $OUTPUT_DIR -o $OUTPUT_DIR -n raw_reads_multiqc\n",
    "\n",
    "echo \"Finished raw reads multiqc report.\""
   ]
  },
  {
   "cell_type": "raw",
   "id": "0ced72eb-8635-463f-b5c3-982d07a81208",
   "metadata": {},
   "source": [
    "01_trim_fastq.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e78512b4-fcbc-4fb5-9e4f-85a6308916c5",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=trim_fastq\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH --mem=8GB\n",
    "#SBATCH --array=01-25\n",
    "#SBATCH --partition=large\n",
    "#SBATCH --output=trim_fastq.%A_%a.out # Include the job ID in the names of the\n",
    "#SBATCH --error=trim_fastq.%A_%a.err # Output and error files\n",
    "\n",
    "# load modules \n",
    "module load TrimGalore/0.6.10-gimkl-2022a-Python-3.11.3-Perl-5.34.1\n",
    "\n",
    "# define variables\n",
    "RAW_FASTQ_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/trimmed_files/\"\n",
    "\n",
    "mkdir -p $OUTPUT_DIR\n",
    "\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# trim\n",
    "# -q 20: This option specifies the minimum quality score for trimming. \n",
    "# Trim Galore! will remove low-quality bases from the 3' end of reads until a base with a quality score of 20 or higher is encountered.\n",
    "# trim for each lane because there may be variations in optical duplicate production from each sequencing lane\n",
    "for lane in {1..4}; do\n",
    "    trim_galore -q 20 -o $OUTPUT_DIR/ --paired $RAW_FASTQ_DIR/AC${SAMPLE_ID}_S*_L00${lane}_R1_001.fastq.gz \\\n",
    "        $RAW_FASTQ_DIR/AC${SAMPLE_ID}_S*_L00${lane}_R2_001.fastq.gz \n",
    "done\n",
    "\n",
    "echo \"Trimming completed for sample ${SAMPLE_ID} in lane ${lane}\""
   ]
  },
  {
   "cell_type": "raw",
   "id": "64092e13-fb5d-4c9e-8716-48be0aff13ae",
   "metadata": {},
   "source": [
    "02_posttrim_fastqc.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b732754d-b1d0-4e7c-b5c9-3e4492034a9a",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=post_trim_fastqc\n",
    "#SBATCH --array=01-25\n",
    "#SBATCH --time=01:30:00\n",
    "#SBATCH --mem=1GB\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --output=post_trim_fastqc_%A_%a.out\n",
    "#SBATCH --error=post_trim_fastqc_%A_%a.err\n",
    "\n",
    "#load modules\n",
    "module load FastQC/0.12.1\n",
    "\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# ppaths\n",
    "TRIMMED_FASTQ_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/trimmed_files/\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/trimmed_fastqc_results\"\n",
    "\n",
    "mkdir -p $OUTPUT_DIR\n",
    "\n",
    "# run fastqc\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# Run FastQC on R1 and R2 files\n",
    "for lane in {1..4}; do\n",
    "    fastqc $TRIMMED_FASTQ_DIR/AC${SAMPLE_ID}_S*_L00${lane}_R1_001_val_1.fq.gz \\\n",
    "        $TRIMMED_FASTQ_DIR/AC${SAMPLE_ID}_S*_L00${lane}_R2_001_val_2.fq.gz \\\n",
    "        --outdir=$OUTPUT_DIR\n",
    "\n",
    "        echo \"Post trim FastQC analysis completed for sample ${SAMPLE_ID} in lane ${lane}\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "13dcebc8-ba14-4b32-b760-ed1ad5460597",
   "metadata": {},
   "source": [
    "02_posttrim_multiqc.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7a53b0f6-804c-441a-ba63-9a686450e43d",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=post_trim_multiqc\n",
    "#SBATCH --time=00:15:00\n",
    "#SBATCH --mem=3GB\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --output=post_trim_multiqc_%J.out\n",
    "#SBATCH --error=post_trim_multiqc_%J.err\n",
    "\n",
    "#load modules\n",
    "module load MultiQC\n",
    "\n",
    "# ppaths\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/trimmed_fastqc_results\"\n",
    "\n",
    "# run multiqc\n",
    "multiqc $OUTPUT_DIR -o $OUTPUT_DIR -n post_trim_multiqc\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c195c75f-c9e4-42ea-91f9-dca78fe1926f",
   "metadata": {},
   "source": [
    "03a_merge_R1_R2_data.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a41f3072-4334-45ca-80ec-124533498802",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=merge_data\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --time=00:15:00\n",
    "#SBATCH --output=merge_data_%J.out\n",
    "#SBATCH --error=merge_data_%J.err\n",
    "\n",
    "INPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/trimmed_files\n",
    "\n",
    "# merge the data from each lane into single files for R1 and r2 reads\n",
    "\n",
    "cd $INPUT_DIR\n",
    "\n",
    "# r1\n",
    "for i in {01..25}; do\n",
    "    # remove leading zeros and convert to integer\n",
    "    i_int=$((10#$i))\n",
    "\n",
    "    num=$((49 + i_int))  # For i=01 -> num=50, i=02 -> num=51, etc.\n",
    "\n",
    "    echo \"Merging lanes 1-4 for index AC${i}_S${num}\"\n",
    "\n",
    "    # Concatenate files from lanes 1-4 and compress the output\n",
    "    cat AC${i}_S${num}_L001_R1_001_val_1.fq.gz \\\n",
    "        AC${i}_S${num}_L002_R1_001_val_1.fq.gz \\\n",
    "        AC${i}_S${num}_L003_R1_001_val_1.fq.gz \\\n",
    "        AC${i}_S${num}_L004_R1_001_val_1.fq.gz \\\n",
    "        > AC${i}_merged_R1.fq.gz\n",
    "\techo \"Merged R1 files for sample ${i}\"\n",
    "done\n",
    "\n",
    "# r2\n",
    "for i in {01..25}; do\n",
    "    i_int=$((10#$i))\n",
    "    num=$((49 + i_int))  # For i=01 -> num=50, i=02 -> num=51, etc.\n",
    "    \n",
    "    echo \"Merging lanes 1-4 for index A${i}_S${num}\"\n",
    "    \n",
    "    # Concatenate files from lanes 1-4 and compress the output\n",
    "    cat AC${i}_S${num}_L001_R2_001_val_2.fq.gz \\\n",
    "        AC${i}_S${num}_L002_R2_001_val_2.fq.gz \\\n",
    "        AC${i}_S${num}_L003_R2_001_val_2.fq.gz \\\n",
    "        AC${i}_S${num}_L004_R2_001_val_2.fq.gz \\\n",
    "        > AC${i}_merged_R2.fq.gz\n",
    "\techo \"Merged R2 files for sample ${i}\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "2ddd6548-88de-4ea9-b1fc-d2202c44bbd8",
   "metadata": {},
   "source": [
    "03_map_reads.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "f397b3b2-b8fd-44ca-a48e-709e9826a288",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=bwa_mapping\n",
    "#SBATCH --array=01-25\n",
    "#SBATCH --time=08:00:00\n",
    "#SBATCH --mem=32GB\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH --output=bwa_mapping_%A_%a.out\n",
    "#SBATCH --error=bwa_mapping_%A_%a.err\n",
    "\n",
    "#load modules\n",
    "module load BWA/0.7.17-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set variables\n",
    "INPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/trimmed_files\n",
    "REFERENCE_GENOME=/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\n",
    "OUTPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\n",
    "\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# if you get bwa index error messages, ensure you have generated the bwa index files and store them in the same directory as your ref seq\n",
    "\n",
    "# use bwa mem to align paired end reads to the reference genome\n",
    "bwa mem -t 32 $REFERENCE_GENOME $INPUT_DIR/AC${SAMPLE_ID}_merged_R1.fq.gz $INPUT_DIR/AC${SAMPLE_ID}_merged_R2.fq.gz \\\n",
    "    > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment.sam\n",
    "\n",
    "echo \"Finished alignments.\"\n",
    "\n",
    "# convert SAM to BAM and sort. Filter out reads with MAPQUALITY less than 30 and unmapped reads.\n",
    "samtools view -q 30 -F 4 -@ 32 -b $OUTPUT_DIR/AC${SAMPLE_ID}_alignment.sam | samtools sort -@ 32 -o $OUTPUT_DIR/AC${SAMPLE_ID}_alignment.sorted.bam\n",
    "samtools index -@ 32 $OUTPUT_DIR/AC${SAMPLE_ID}_alignment.sorted.bam\n",
    "\n",
    "echo \"Finished sorting and indexing mapped reads.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a37c7d05-050c-4647-8fcf-62f31e5af521",
   "metadata": {},
   "source": [
    "# Dunedin samples, 03_map_reads.sh (pre-mapping  (e.g., trimming done by Brooke, fastqcs show trimmed samples and all samples passing qual filters)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "98f2a006-6b58-48f3-a4ca-2452dfa72a6a",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=bwa_mapping\n",
    "#SBATCH --array=1-8\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH --mem=36GB\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --output=bwa_mapping_%A_%a.out\n",
    "#SBATCH --error=bwa_mapping_%A_%a.err\n",
    "\n",
    "#load modules\n",
    "module load BWA/0.7.17-GCC-11.3.0\n",
    "module load SAMtools/1.19-GCC-12.3.0\n",
    "\n",
    "# set variables\n",
    "INPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/dunedin_samples\n",
    "REFERENCE_GENOME=/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\n",
    "OUTPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/dunedin_samples/alignments\n",
    "\n",
    "SAMPLE_ID=$(sed -n \"$((SLURM_ARRAY_TASK_ID))p\" filenames.txt)\n",
    "\n",
    "FWD_FILENAME=\"CSD_0${SAMPLE_ID}_R1_T.fastq.gz\"\n",
    "RVS_FILENAME=\"CSD_0${SAMPLE_ID}_R2_T.fastq.gz\"\n",
    "\n",
    "echo \"Starting BWA alignment for $FWD_FILENAME and $RVS_FILENAME.\"\n",
    "\n",
    "# use bwa mem to align paired end reads to the reference genome\n",
    "bwa mem -t 16 $REFERENCE_GENOME $FWD_FILENAME $RVS_FILENAME \\\n",
    "    > $OUTPUT_DIR/CSD_0${SAMPLE_ID}_alignment.sam\n",
    "\n",
    "echo \"Finished alignments.\"\n",
    "\n",
    "# convert SAM to BAM and sort. Filter out reads with MAPQUALITY less than 30 and unmapped reads.\n",
    "samtools view -q 30 -F 4 -@ 16 -b $OUTPUT_DIR/CSD_0${SAMPLE_ID}_alignment.sam | samtools sort -@ 16 -o $OUTPUT_DIR/CSD_0${SAMPLE_ID}_alignment.sorted.bam\n",
    "samtools index -@ 16 $OUTPUT_DIR/CSD_0${SAMPLE_ID}_alignment.sorted.bam\n",
    "\n",
    "echo \"Finished sorting and indexing mapped reads for sample $SAMPLE_ID.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "68b311eb-c31d-46a2-9c8b-1110ef641f78",
   "metadata": {},
   "source": [
    "03_addRG_mapped_reads.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6fec2482-8df2-4b62-bd8b-48ffb8ce63c6",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=addRG\n",
    "#SBATCH --time=01:00:00\n",
    "#SBATCH --array=01-33\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --output=addRG_%A_%a.out\n",
    "#SBATCH --error=addRG_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load modules\n",
    "module load GATK/4.5.0.0-gimkl-2022a\n",
    "\n",
    "# set variables\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "BAM_FILELIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/bam_filelist.txt\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "BAM_FILE=$(sed -n \"${ARRAY_INDEX}p\" \"$BAM_FILELIST\")\n",
    "\n",
    "# extract the basename\n",
    "BASE_NAME=$(basename \"$BAM_FILE\" _alignment.sorted.bam)\n",
    "\n",
    "echo \"Processing sample $BASE_NAME.\"\n",
    "\n",
    "# gatk add read groups\n",
    "gatk AddOrReplaceReadGroups \\\n",
    "    I=\"$BAM_FILE\" \\\n",
    "    O=\"$OUTPUT_DIR/${BASE_NAME}_alignment.sorted_RG.bam\" \\\n",
    "    RGID=\"$BASE_NAME\" \\\n",
    "    RGLB=UnknownLibrary \\\n",
    "    RGPL=ILLUMINA \\\n",
    "    RGPU=unit1 \\\n",
    "    RGSM=\"$BASE_NAME\"\n",
    "\n",
    "echo \"Finished adding read groups.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1163ec20-8fc0-4917-814e-7e901df9922c",
   "metadata": {},
   "source": [
    "03_remove_dups.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "20c88149-c9fa-48f8-9ef9-c8a49c22af7b",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=rm_dups\n",
    "#SBATCH --time=01:30:00\n",
    "#SBATCH --array=11-33\n",
    "#SBATCH --mem=72GB\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --output=rm_dups_%A_%a.out\n",
    "#SBATCH --error=rm_dups_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load modules\n",
    "module load GATK/4.5.0.0-gimkl-2022a\n",
    "\n",
    "# note that samples AC 1-10 were run with 48Gb before the rest were run with increased memory\n",
    "\n",
    "# set variables\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "BAM_FILELIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/bam_filelist.txt\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "BAM_FILE=$(sed -n \"${ARRAY_INDEX}p\" \"$BAM_FILELIST\")\n",
    "\n",
    "# extract the basename\n",
    "BASE_NAME=$(basename \"$BAM_FILE\" _alignment.sorted.bam)\n",
    "\n",
    "echo \"Processing sample $BASE_NAME.\"\n",
    "\n",
    "# filter duplicates \n",
    "gatk --java-options \"-Xmx48g\" MarkDuplicates \\\n",
    "    -I \"$OUTPUT_DIR/${BASE_NAME}_alignment.sorted_RG.bam\" \\\n",
    "    -O \"$OUTPUT_DIR/${BASE_NAME}_alignment.sorted_RGMD.bam\" \\\n",
    "    -M \"$OUTPUT_DIR/${BASE_NAME}_alignment.sorted_RGMD_metrics.txt\" \\\n",
    "    --REMOVE_DUPLICATES\n",
    "\n",
    "echo \"Finished processing duplicates.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "16d67b62-056a-4245-b6bb-bcc63dd1e601",
   "metadata": {},
   "source": [
    "03_index_RGMD.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9ccb6e7a-6f52-456e-81b0-36561c18c130",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=index_RGMD\n",
    "#SBATCH --time=02:00:00\n",
    "#SBATCH --array=01-33\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --output=index_RGMD_%A_%a.out\n",
    "#SBATCH --error=index_RGMD_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load modules\n",
    "module load SAMtools\n",
    "\n",
    "# set variables\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "BAM_FILELIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/RGMD_bam_filelist.txt\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "BAM_FILE=$(sed -n \"${ARRAY_INDEX}p\" \"$BAM_FILELIST\")\n",
    "\n",
    "# extract the basename\n",
    "BASE_NAME=$(basename \"$BAM_FILE\" _alignment.sorted_RGMD.bam)\n",
    "\n",
    "echo \"Processing sample $BASE_NAME.\"\n",
    "\n",
    "# index the RGMD bams\n",
    "samtools index $BAM_FILE\n",
    "\n",
    "echo \"Finished indexing $BASE_NAME.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "efc96838-4525-4d65-aac4-490e9a621e54",
   "metadata": {},
   "source": [
    "03_calc_depth.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "79927bdf-ac8c-43db-8b0b-8ceb1e382f76",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=calc_depth\n",
    "#SBATCH --time=06:00:00\n",
    "#SBATCH --array=01-33\n",
    "#SBATCH --mem=1GB\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --output=calc_depth_%A_%a.out\n",
    "#SBATCH --error=calc_depth_%A_%a.err\n",
    "\n",
    "#load modules\n",
    "module load SAMtools\n",
    "\n",
    "# set variables\n",
    "BAM_FILELIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/RGMD_bam_filelist.txt\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "BAM_FILE=$(sed -n \"${ARRAY_INDEX}p\" \"$BAM_FILELIST\")\n",
    "\n",
    "# extract the basename\n",
    "BASE_NAME=$(basename \"$BAM_FILE\" _alignment.sorted_RGMD.bam)\n",
    "\n",
    "echo \"Processing sample $BASE_NAME.\"\n",
    "\n",
    "# calculate coverage\n",
    "samtools depth -a $BAM_FILE > $OUTPUT_DIR/${BASE_NAME}_alignment.sorted_RGMD.coverage.txt\n",
    "\n",
    "echo \"Completed processing sample $BASE_NAME.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "63ba4cc7-1104-4410-beed-c95761c438fd",
   "metadata": {},
   "source": [
    "03_calc_cov.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7ffe37a2-e7d6-410c-8714-cfe1ac1f8ba7",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=calc_cov\n",
    "#SBATCH --time=01:00:00\n",
    "#SBATCH --array=01-33\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --output=calc_cov_%A_%a.out\n",
    "#SBATCH --error=calc_cov_%A_%a.err\n",
    "\n",
    "# Load the necessary modules\n",
    "module load Python\n",
    "\n",
    "# Directory containing the coverage files\n",
    "FILE_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/coverage_file_list.txt\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "FILE=$(sed -n \"${ARRAY_INDEX}p\" \"$FILE_LIST\")\n",
    "\n",
    "# extract the basename\n",
    "BASE_NAME=$(basename \"$FILE\" _alignment.sorted_RGMD.coverage.txt)\n",
    "\n",
    "echo \"Processing sample $BASE_NAME.\"\n",
    "\n",
    "# Run the Python script\n",
    "python 03_calculate_average_coverage.py $FILE\n",
    "\n",
    "# then you'll need to cat the contents of the output files to a coverage_results.txt file\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7cb6ca3e-6edb-44e3-a079-4de71eb5382b",
   "metadata": {},
   "source": [
    "03_calculate_average_coverage.py"
   ]
  },
  {
   "cell_type": "raw",
   "id": "42f2dde0-b097-4bb2-9cb9-6b640c7f48a3",
   "metadata": {},
   "source": [
    "import sys\n",
    "\n",
    "def calculate_average_coverage(cov_file):\n",
    "    total_coverage = 0\n",
    "    total_positions = 0\n",
    "\n",
    "    with open(cov_file, 'r') as file:\n",
    "        for line in file:\n",
    "            _, _, coverage = line.strip().split()\n",
    "            total_coverage += int(coverage)\n",
    "            total_positions += 1\n",
    "\n",
    "    average_coverage = total_coverage / total_positions if total_positions > 0 else 0\n",
    "    return average_coverage\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    if len(sys.argv) != 2:\n",
    "        print(\"Usage: python calculate_average_coverage.py <coverage.txt>\")\n",
    "        sys.exit(1)\n",
    "\n",
    "    cov_file = sys.argv[1]\n",
    "    average_coverage = calculate_average_coverage(cov_file)\n",
    "    suffix_string = \"_alignment.sorted_RGMD.coverage.txt\"\n",
    "    trimmed_cov_file = cov_file.replace(suffix_string, \"\")\n",
    "\n",
    "    print(f\"{trimmed_cov_file}\\t{average_coverage}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "966994a7-e4ed-46c7-afec-4cf1d22ee25e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Mitogenome section"
   ]
  },
  {
   "cell_type": "raw",
   "id": "11fb126a-d24b-4f1c-86f0-c8323046ae63",
   "metadata": {},
   "source": [
    "03_map_to_mitogenomes.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1dd0b192-5c5f-4b40-a9b0-aaeee35a7635",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=bwa_mitogenome\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --array=01-25\n",
    "#SBATCH --mem=24GB\n",
    "#SBATCH --time=36:00:00\n",
    "#SBATCH --output=bwa_mitogenome_%A_%a.out\n",
    "#SBATCH --error=bwa_mitogenome_%A_%a.err\n",
    "\n",
    "module load picard\n",
    "module load BWA\n",
    "module load SAMtools\n",
    "\n",
    "# set variables\n",
    "INPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/trimmed_files\n",
    "REFERENCE_GENOME=/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/Rattus_mito_index/Rattus_mitogenome_NCBI.fa\n",
    "OUTPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/mitogenome_mapping\n",
    "\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# use bwa mem to align paired end reads to the reference genome\n",
    "bwa mem -t 16 $REFERENCE_GENOME $INPUT_DIR/AC${SAMPLE_ID}_merged_R1.fq.gz $INPUT_DIR/AC${SAMPLE_ID}_merged_R2.fq.gz \\\n",
    "    > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sam\n",
    "\n",
    "echo \"BWA alignment complete for ${SAMPLE_ID}.\"\n",
    "\n",
    "# convert SAM to BAM and sort\n",
    "samtools view -@ 16 -bS $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sam | samtools sort -@ 16 -o $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.bam\n",
    "\n",
    "#mark duplicates\n",
    "picard MarkDuplicates -I $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.bam \\\n",
    "    -M $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates_duplicate_metrics.txt \\\n",
    "    -O $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam \\\n",
    "    --OPTICAL_DUPLICATE_PIXEL_DISTANCE 250 \\\n",
    "    --VALIDATION_STRINGENCY SILENT \\\n",
    "    -MAX_FILE_HANDLES 1000 \\\n",
    "    --MAX_RECORDS_IN_RAM 5000\n",
    "\n",
    "# generate intitial summary stats for the marked duplicates alignment\n",
    "samtools flagstat $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.flagstat.txt\n",
    "\n",
    "echo \"samtools flagstat complete for ${SAMPLE_ID}.\"\n",
    "\n",
    "# index sorted BAM file\n",
    "samtools index $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam\n",
    "\n",
    "# remove duplicates for samtools depth calcs, remove duplicates (flagged as quality scores of 1024)\n",
    "samtools view -b -F 1024 $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.bam \n",
    "\n",
    "# index the deduplicated file\n",
    "samtools index $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.bam\n",
    "\n",
    "# calculate coverage\n",
    "samtools depth -a $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.bam > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.coverage.txt\n",
    "\n",
    "# cleanup\n",
    "rm $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sam\n",
    "\n",
    "echo \"BWA mapping and SAM/BAM processing completed.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "565db819-252d-46d7-90af-65cd395abce9",
   "metadata": {},
   "source": [
    "03_mitogenome_stats.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "3cf7d95d-f0f3-4bf3-83b5-4d3958a86322",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --job-name=mitogenome_stats\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --array=01-25\n",
    "#SBATCH --mem=6GB\n",
    "#SBATCH --time=01:00:00\n",
    "#SBATCH --output=bwa_mitogenome_%A_%a.out\n",
    "#SBATCH --error=bwa_mitogenome_%A_%a.err\n",
    "\n",
    "module load picard\n",
    "module load SAMtools\n",
    "\n",
    "# set variables\n",
    "INPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/trimmed_files\n",
    "REFERENCE_GENOME=/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/Rattus_mito_index/Rattus_mitogenome_NCBI.fa\n",
    "OUTPUT_DIR=/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/mitogenome_mapping\n",
    "\n",
    "# sample id, make sure you are including the prceding 0 in the numbers leading up to 10.\n",
    "SAMPLE_ID=$(printf \"%02d\" $SLURM_ARRAY_TASK_ID)\n",
    "\n",
    "# use bwa mem to align paired end reads to the reference genome\n",
    "bwa mem -t 16 $REFERENCE_GENOME $INPUT_DIR/AC${SAMPLE_ID}_merged_R1.fq.gz $INPUT_DIR/AC${SAMPLE_ID}_merged_R2.fq.gz \\\n",
    "    > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sam\n",
    "\n",
    "echo \"BWA alignment complete for ${SAMPLE_ID}.\"\n",
    "\n",
    "# convert SAM to BAM and sort\n",
    "samtools view -@ 16 -bS $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sam | samtools sort -@ 16 -o $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.bam\n",
    "\n",
    "#mark duplicates\n",
    "picard MarkDuplicates -I $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.bam \\\n",
    "    -M $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates_duplicate_metrics.txt \\\n",
    "    -O $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam \\\n",
    "    --OPTICAL_DUPLICATE_PIXEL_DISTANCE 250 \\\n",
    "    --VALIDATION_STRINGENCY SILENT \\\n",
    "    -MAX_FILE_HANDLES 1000 \\\n",
    "    --MAX_RECORDS_IN_RAM 5000\n",
    "\n",
    "# generate intitial summary stats for the marked duplicates alignment\n",
    "samtools flagstat $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.flagstat.txt\n",
    "\n",
    "echo \"samtools flagstat complete for ${SAMPLE_ID}.\"\n",
    "\n",
    "# index sorted BAM file\n",
    "samtools index $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam\n",
    "\n",
    "# remove duplicates for samtools depth calcs, remove duplicates (flagged as quality scores of 1024)\n",
    "samtools view -b -F 1024 $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.markedduplicates.bam > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.bam\n",
    "\n",
    "# index the deduplicated file\n",
    "samtools index $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.bam\n",
    "\n",
    "# calculate coverage\n",
    "samtools depth -a $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.bam > $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sorted.deduplicated.coverage.txt\n",
    "\n",
    "# cleanup\n",
    "rm $OUTPUT_DIR/AC${SAMPLE_ID}_alignment_mitogenome.sam\n",
    "\n",
    "echo \"BWA mapping and SAM/BAM processing completed.\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cf4726a-bb50-47aa-980d-0ba4edd41563",
   "metadata": {},
   "source": [
    "------ back to Nuclear genome analysis---"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9a662ea2-2c0f-47e1-9349-a60b260c4a2f",
   "metadata": {},
   "source": [
    "05_Ychr_test.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9c442152-d881-42e3-ba3f-03de5a2d4764",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=SRY_test\n",
    "#SBATCH --time=00:15:00\n",
    "#SBATCH --mem=512MB\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --output=SRY_test_%A.out\n",
    "#SBATCH --error=SRY_test_%A.err\n",
    "\n",
    "#load modules\n",
    "module load SAMtools\n",
    "\n",
    "# Directory containing BAM files\n",
    "BAM_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "# Output file to store results\n",
    "OUTPUT_FILE=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/SRY_read_counts.txt\"\n",
    "\n",
    "# Create or clear the output file\n",
    "echo \"Sample Name, Total Reads Mapped to SRY-protein-like(NC_046173.1:1926850-1927383)\" > $OUTPUT_FILE\n",
    "\n",
    "# Loop over each BAM file in the directory\n",
    "for bam_file in \"$BAM_DIR\"/*_RGMD.bam; do\n",
    "  # Get the sample name from the BAM file name\n",
    "  sample_name=$(basename \"$bam_file\" _alignment.sorted_RGMD.bam)\n",
    "\n",
    "  # Extract the number of reads mapped to the SRY region of the Y chromosome\n",
    "  sry_reads_count=$(samtools view -c \"$bam_file\" \"NC_046173.1:1926850-1927383\")\n",
    "\n",
    "  # Write the result to the output file\n",
    "  echo \"$sample_name, $sry_reads_count\" >> $OUTPUT_FILE\n",
    "done\n",
    "\n",
    "echo \"SRY read counts analysis completed. Results are in $OUTPUT_FILE.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "1fa8e9f8-0346-459d-ada5-89a93be21fc5",
   "metadata": {},
   "source": [
    "06_submit_gatk_arrays.sh\n",
    "# get the alignment file for each sample"
   ]
  },
  {
   "cell_type": "raw",
   "id": "810d44fe-ff84-423a-a129-9397fcdf96e1",
   "metadata": {},
   "source": [
    "BAM_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/bamRGMD_filelist.txt\"\n",
    "\n",
    "for BAM_FILE in $(cat $BAM_LIST); do\n",
    "\tsbatch 06_call_vars_gatk.sh $BAM_FILE\n",
    "done\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9ad36f76-126a-471d-8a31-bf92cb49bbc1",
   "metadata": {},
   "source": [
    "06_call_vars_gatk.sh\n",
    "# run an array - split sample by chromosome (i.e.interval list) and produce a gVCF from each chr from each sample"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9a37516b-0443-4bce-a3f8-1ecfd7b930f5",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=call_vars_GATK\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH --mem=24GB\n",
    "#SBATCH --array=01-21\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH --output=call_vars_GATK_%A_%a.out\n",
    "#SBATCH --error=call_vars_GATK_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load mods\n",
    "module load GATK\n",
    "\n",
    "# set args\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "INTERVAL_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/interval.list\"\n",
    "\n",
    "BAM_FILE=$1\n",
    "\n",
    "# extract the basename and make a sample directory\n",
    "BASE_NAME=$(basename \"$BAM_FILE\" _alignment.sorted_RGMD.bam)\n",
    "\n",
    "mkdir -p $OUTPUT_DIR/${BASE_NAME}\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# assign each chromosome to an array\n",
    "CHR=$(sed -n \"${ARRAY_INDEX}p\" \"$INTERVAL_LIST\")\n",
    "\n",
    "echo \"Processing sample: ${BASE_NAME}, chromosome: ${CHR}.\"\n",
    "\n",
    "# create the .dict file for the ref fasta for gatk\n",
    "# gatk --java-options \"-Xmx24g -XX:ParallelGCThreads=16\" CreateSequenceDictionary \\\n",
    "#    -R $REFERENCE_GENOME\n",
    "\n",
    "# Run HaplotypeCaller to generate gVCF\n",
    "gatk --java-options \"-Xmx24g -XX:ParallelGCThreads=16\" HaplotypeCaller \\\n",
    "\t-R $REFERENCE_GENOME \\\n",
    "\t-I $BAM_FILE \\\n",
    "\t-O $OUTPUT_DIR/${BASE_NAME}/${BASE_NAME}_${CHR}.g.vcf.gz \\\n",
    "\t-ERC GVCF \\\n",
    "\t-L $CHR\n",
    "\n",
    "echo \"Haplotyping gVCF completed for sample $BASE_NAME, chr: $CHR. Files in $OUTPUT_DIR/${BASE_NAME}.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a3198342-4123-4023-ae5f-10c7c8486d16",
   "metadata": {},
   "source": [
    "06_importDB_GATK.sh\n",
    "# create a database for each chromosome"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d4ac4ad2-6de0-467b-8cd5-aeb5968c66ce",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=importDB_GATK\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH --mem=8GB\n",
    "#SBATCH --array=01-21\n",
    "#SBATCH --cpus-per-task=8    \n",
    "#SBATCH --output=importDB_GATK_%A_%a.out\n",
    "#SBATCH --error=importDB_GATK_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load mods\n",
    "module load GATK\n",
    "\n",
    "# set args\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "INTERVAL_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/interval.list\"\n",
    "SAMPLE_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/sample_list.txt\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "CHR=$(sed -n \"${ARRAY_INDEX}p\" \"$INTERVAL_LIST\")\n",
    "\n",
    "echo \"Creating genome database for chromosome: $CHR\"\n",
    "\n",
    "# make sure you leave some surplus memory for the native code to use\n",
    "gatk --java-options \"-Xmx6g\" GenomicsDBImport \\\n",
    "    -R $REFERENCE_GENOME \\\n",
    "    --genomicsdb-workspace-path ${OUTPUT_DIR}/db_chr${CHR} \\\n",
    "    --sample-name-map $OUTPUT_DIR/sample_map_chr${CHR}.txt \\\n",
    "    --intervals $CHR\n",
    "\n",
    "echo \"Database generated for ${CHR} in ${OUTPUT_DIR}/db_chr${CHR}\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d646f782-c0d7-48dc-b429-a6f57c445ebd",
   "metadata": {},
   "source": [
    "06_genotypegVCF.sh\n",
    "# joint genotype each chromosome (all samples)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ff49d2cc-5d08-47dc-b938-d2d02f1bb255",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=genotypeGVCF_GATK\n",
    "#SBATCH --time=36:00:00\n",
    "#SBATCH --mem=6GB\n",
    "#SBATCH --array=01-21\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --output=genotypeGVCF_GATK_%A_%a.out\n",
    "#SBATCH --error=genotypeGVCF_GATK_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load mods\n",
    "module load GATK\n",
    "\n",
    "# set args\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "INTERVAL_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/interval.list\"\n",
    "SAMPLE_LIST=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/sample_list.txt\"\n",
    "\n",
    "# get the current array index\n",
    "ARRAY_INDEX=${SLURM_ARRAY_TASK_ID}\n",
    "\n",
    "# get the absolute path for the bam\n",
    "CHR=$(sed -n \"${ARRAY_INDEX}p\" \"$INTERVAL_LIST\")\n",
    "\n",
    "echo \"Beginning joint genotyping for chromosome: $CHR\"\n",
    "\n",
    "# make sure you leave some surplus memory for the native code to use\n",
    "gatk --java-options \"-Xmx4g -XX:ParallelGCThreads=4\" GenotypeGVCFs \\\n",
    "    -R $REFERENCE_GENOME \\\n",
    "    -V gendb://${OUTPUT_DIR}/db_chr${CHR} \\\n",
    "    -O ${OUTPUT_DIR}/${CHR}_multisample.vcf.gz\n",
    "\n",
    "echo \"Joint genotyping complete for ${CHR}. VCF located in ${OUTPUT_DIR}.\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "cd6f0e03-145e-4c5c-915d-70cbb11e20cb",
   "metadata": {},
   "source": [
    "06_gatherVCFs.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "0bba9ef7-3695-4c9e-96a2-005e5e1ff43e",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=gatherVCF_GATK\n",
    "#SBATCH --time=24:00:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --output=gatherVCF_GATK_%A_%a.out\n",
    "#SBATCH --error=gatherVCF_GATK_%A_%a.err\n",
    "#SBATCH --mail-type=END,FAIL\n",
    "#SBATCH --mail-user=anna.clark@postgrad.otago.ac.nz\n",
    "\n",
    "#load mods\n",
    "module load picard\n",
    "\n",
    "# set args\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "\n",
    "# you need to pass tthe JVM options directly rather than using --java-options flag\n",
    "export JAVA_OPTS=\"-Xmx3g -XX:ParallelGCThreads=6\"\n",
    "\n",
    "picard GatherVcfs \\\n",
    "   -I ${OUTPUT_DIR}/NC_046154.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046155.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046156.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046157.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046158.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046159.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046160.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046161.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046162.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046163.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046164.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046165.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046166.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046167.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046168.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046169.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046170.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046171.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046172.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_046173.1_multisample.vcf.gz \\\n",
    "   -I ${OUTPUT_DIR}/NC_012374.1_multisample.vcf.gz \\\n",
    "   -O ${OUTPUT_DIR}/multisample_combined.vcf\n",
    "\n",
    "echo \"Scattered chromosome VCF files successfully combined. File located in ${OUTPUT_DIR}.\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b63e7f5-77c5-4381-a220-deb72ddd75ae",
   "metadata": {},
   "source": [
    "- mapping qual: \"MQ < 40.0\" \n",
    "- quality/depth: \"QD < 2.0\"\n",
    "- qual: \"QUAL < 30.0\"\n",
    "- strand odds ratio (strand bias): \"SOR > 3.0\"\n",
    "- Fisher ET for strand bias:  \"FS > 60.0\" \n",
    "- mapping qual rank sum: \"MQRankSum < -12.5\"\n",
    "- read position rank sum: \"ReadPosRankSum < -8.0\"\n",
    "- depth: \"DP < 10\""
   ]
  },
  {
   "cell_type": "raw",
   "id": "f50786db-157a-4e86-af94-22da60d3625c",
   "metadata": {},
   "source": [
    "08_advancedfilter_vars_gatk.sh\n",
    "\n",
    "This depth filter functions on the depth value at the site across all samples. So setting it at 10 makes it obsolete. We need to use BCFtools below to access the depth within each sample's FORMAT field (not INFO, which GATK searches). Set a minimum and a maximum depth."
   ]
  },
  {
   "cell_type": "raw",
   "id": "0e72c4e7-781e-4c5a-b9f1-d6a698e75314",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=advancedvar_filter\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=24GB\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --output=advancedvar_filter_%A.out\n",
    "#SBATCH --error=advancedvar_filter_%A.err\n",
    "\n",
    "#load mods\n",
    "module load GATK\n",
    "\n",
    "# paths\n",
    "REFERENCE_GENOME=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic.fna.gz\"\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "OUTPUT_VCF=\"${OUTPUT_DIR}/multisample_combined.vcf\"\n",
    "OUTPUT_SNP_VCF=\"${OUTPUT_DIR}/advanced_filtering_results/advmultisample_combined_snps.vcf\"\n",
    "OUTPUT_SNP_FILTERED=\"${OUTPUT_DIR}/advanced_filtering_results/advmultisample_combined_snps_filtered.vcf\"\n",
    "\n",
    "######## filtering on the joint genotyped VCF\n",
    "## dont need to normalise this vcf\n",
    "# so we need biallelic snps only, raw read depth more than 10\n",
    "\n",
    "# Select biallelic SNPs\n",
    "gatk --java-options \"-Xmx22g -XX:ParallelGCThreads=24\" SelectVariants \\\n",
    "    -V $OUTPUT_VCF \\\n",
    "    -select-type SNP \\\n",
    "    -restrict-alleles-to BIALLELIC \\\n",
    "    -O $OUTPUT_SNP_VCF\n",
    "\n",
    "echo \"Biallelic SNP selection completed. Output file: $OUTPUT_SNP_VCF\"\n",
    "\n",
    "# now depth and reduce false positives and strand bias in the dataset\n",
    "gatk --java-options \"-Xmx22g -XX:ParallelGCThreads=24\" VariantFiltration \\\n",
    "    -V $OUTPUT_SNP_VCF \\\n",
    "    --filter-expression \"MQ < 40.0\" --filter-name \"MQ40\" \\\n",
    "    --filter-expression \"QD < 2.0\" --filter-name \"QD2\" \\\n",
    "    --filter-expression \"QUAL < 30.0\" --filter-name \"QUAL30\" \\\n",
    "    --filter-expression \"SOR > 3.0\" --filter-name \"SOR3\" \\\n",
    "    --filter-expression \"FS > 60.0\" --filter-name \"FS60\" \\\n",
    "    --filter-expression \"MQRankSum < -12.5\" --filter-name \"MQRankSum-12.5\" \\\n",
    "    --filter-expression \"ReadPosRankSum < -8.0\" --filter-name \"ReadPosRankSum-8\" \\\n",
    "    --filter-expression \"DP < 10\" --filter-name \"LowDP\" \\\n",
    "    --set-filtered-genotype-to-no-call \\\n",
    "    -O $OUTPUT_SNP_FILTERED\n",
    "\n",
    "echo \"Filtering completed. Output file: $OUTPUT_SNP_FILTERED\"\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "e0947bc0-4afe-4d33-b51b-b05727fd73d4",
   "metadata": {},
   "source": [
    "09_bcftools_qualitysnps.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7fe3ec77-3bc8-4c0d-96f1-c0958c0a5bbd",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=bcftools_qualsnps\n",
    "#SBATCH --time=03:00:00\n",
    "#SBATCH --mem=24GB\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH --output=bcftools_qualsnps_%A.out\n",
    "#SBATCH --error=bcftools_qualsnps_%A.err\n",
    "\n",
    "#load mods\n",
    "module load BCFtools\n",
    "\n",
    "# paths\n",
    "OUTPUT_DIR=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments\"\n",
    "OUTPUT_SNP_FILTERED=\"${OUTPUT_DIR}/advanced_filtering_results/advmultisample_combined_snps_filtered.vcf\"\n",
    "QUAL_SNPS=\"${OUTPUT_DIR}/advanced_filtering_results/advmultisample_combined_snps_filtered_passonly.vcf\"\n",
    "\n",
    "# keep only the snps that have passed out GATK filters (i.e., FILTER=PASS)\n",
    "bcftools view -f PASS ${OUTPUT_SNP_FILTERED} > ${QUAL_SNPS}\n",
    "\n",
    "echo \"Keeping only variants with filter=PASS. Output file: $QUAL_SNPS\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c82d1418-da2e-4409-ac8f-6826df5670b5",
   "metadata": {},
   "source": [
    "10_CHRname_map.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "358bc74e-9060-4ad3-a8db-d0e07e636b35",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=chrname_map\n",
    "#SBATCH --time=00:15:00\n",
    "#SBATCH --mem=8GB\n",
    "#SBATCH --qos=debug\n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --output=chrname_map_%A.out\n",
    "#SBATCH --error=chrname_map_%A.err\n",
    "\n",
    "#load modules\n",
    "module load PLINK/1.09b6.16\n",
    "module load BCFtools\n",
    "\n",
    "# dirs\n",
    "WD=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/advanced_filtering_results\"\n",
    "VCF=\"${WD}/advmultisample_combined_snps_filtered_passonly_namesedit.vcf\"\n",
    "VCF2=\"${WD}/advmultisample_combined_snps_filtered_passonly_nameschrsedit.vcf\"\n",
    "CHR_MAP=\"${WD}/chr_map.txt\"\n",
    "\n",
    "# we need to rename the chromosome contigs for plink\n",
    "bcftools annotate --rename-chrs $CHR_MAP -o $VCF2 $VCF"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b6bf46f0-7245-4bbb-ab41-6c7186f2a3c3",
   "metadata": {},
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=uoo03627\n",
    "#SBATCH --job-name=checkdups\n",
    "#SBATCH --time=00:15:00\n",
    "#SBATCH --mem=8GB\n",
    "#SBATCH --qos=debug\n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --output=checkdups_%A.out\n",
    "#SBATCH --error=checkdups_%A.err\n",
    "\n",
    "#load modules\n",
    "module load BCFtools\n",
    "\n",
    "# dirs\n",
    "WD=\"/nesi/nobackup/uoo03627/qt_rat_sequencing/qt_rats_WG/results/reference_genome_alignments/advanced_filtering_results\"\n",
    "pruned_BED=\"${WD}/advmultisample_combined_snps_filtered_minmissing_autosomal_nosingletons_100kbpruned\"\n",
    "maf_BED1=\"${WD}/advmultisample_combined_snps_filtered_minmissing_autosomal_nosingletons_100kbpruned_maf0.05\"\n",
    "maf_BED2=\"${WD}/advmultisample_combined_snps_filtered_minmissing_autosomal_nosingletons_100kbpruned_maf0.1\"\n",
    "pca1=\"${WD}/advmultisample_combined_snps_filtered_minmissing_autosomal_nosingletons_100kbpruned_maf0.05_pca\"\n",
    "pca2=\"${WD}/advmultisample_combined_snps_filtered_minmissing_autosomal_nosingletons_100kbpruned_maf0.1_pca\"\n",
    "\n",
    "# getting rid of any duplicate entries\n",
    "bcftools norm -d none \"${pruned_BED}.vcf\" -o \"${pruned_BED}_dupsrmvd.vcf\"  \n",
    "\n",
    "bcftools norm -d none \"${maf_BED1}.vcf\" -o \"${maf_BED1}_dupsrmvd.vcf\"\n",
    "\n",
    "bcftools norm -d none \"${maf_BED2}.vcf\"  -o \"${maf_BED2}_dupsrmvd.vcf\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 (gimkl-2022a)",
   "language": "python",
   "name": "python3105-gimkl-2022a"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
