{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "85b1955b-8327-44d6-9c2a-a67c37e92e5f",
   "metadata": {},
   "source": [
    "# NgsRelate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abe6a527-a106-4d8b-9cff-3c351389b674",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --job-name=ngsrelate\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH --mem=150GB\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH --output=%x_%A.out\n",
    "#SBATCH --error=%x_%A.err\n",
    "\n",
    "module load Miniconda3/23.10.0-1\n",
    "module load angsd/0.935-GCC-9.2.0\n",
    "module load bzip2/1.0.8-GCCcore-12.3.0\n",
    "\n",
    "export PATH=$PATH:/$workdir/ngsRelate\n",
    "# may also need to update PATH for the newest GLIBXX package\n",
    "\n",
    "workdir=\"/$workdir/results/06_grouped\"\n",
    "outdir=\"/$workdir/results/06_angsd/ngsrelate\"\n",
    "OUTPUT=\"${outdir}/angsd_GL_for_ngsrelate\"\n",
    "BAM_LIST=\"${workdir}/med_cov_rmrelated.bam.filelist\"\n",
    "REF=\"/$workdir/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic_autosomesonly.fna.gz\"\n",
    "FREQ=\"${OUTPUT}.freq\"\n",
    "NGSRELATE_OUTPUT=\"${outdir}/ngsrelate_output\"\n",
    "GL=\"${outdir}/angsd_GL_for_ngsrelate.glf.gz\"\n",
    "\n",
    "mkdir -p $outdir\n",
    "\n",
    "# generate GLs\n",
    "angsd -b $BAM_LIST -gl 2 -domajorminor 1 -snp_pval 1e-6 -domaf 1 -minmaf 0.05 -doGlf 3 -nThreads 8 -out $OUTPUT\n",
    "\n",
    "# extract the frequency column from the allele frequency file and remove the header (required by NgsRelate)\n",
    "zcat \"${OUTPUT}.mafs.gz\" | cut -f5 |sed 1d > $FREQ\n",
    "\n",
    "# run ngsrelate\n",
    "ngsRelate -g $GL -n 41 -f $FREQ -O $NGSRELATE_OUTPUT\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a722986-1d61-484c-9643-c2e6d5d8d982",
   "metadata": {},
   "source": [
    "# PCAngsd\n",
    "Also run with 1% and 5% MAF filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6d5dc3-8b7d-433a-90ac-3af65bbefc7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -l\n",
    "#SBATCH --job-name=geno_angsd\n",
    "#SBATCH --output=%x_%j.out\n",
    "#SBATCH --error=%x_%j.err\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --cpus-per-task=12\n",
    "\n",
    "module load angsd/0.935-GCC-9.2.0\n",
    "\n",
    "workdir=\"/$workdir/results/06_grouped\"\n",
    "outdir=\"/$workdir/results/06_angsd/pcangsd\"\n",
    "BAM_LIST=\"${workdir}/med_cov_rmrelated.bam.filelist\"\n",
    "REF=\"/$workdir/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic_autosomesonly.fna.gz\"\n",
    "\n",
    "mkdir -p $outdir\n",
    "\n",
    "angsd \\\n",
    "    -bam ${BAM_LIST} \\\n",
    "    -ref ${REF} \\\n",
    "    -out \"${outdir}/nomaf\" \\\n",
    "    -nThreads 12 \\\n",
    "    -GL 1 \\\n",
    "    -doGlf 2 \\\n",
    "    -doMaf 2 -doMajorMinor 1 \\\n",
    "    -minMapQ 30 \\\n",
    "    -minQ 20 \\\n",
    "    -SNP_pval 1e-6 \\\n",
    "    -doGeno 2 \\\n",
    "    -doPost 1 -postCutoff 0.95 \\\n",
    "    -minInd 24 \\\n",
    "    -doCounts 1 \\\n",
    "    -setMinDepthInd 4 \\\n",
    "    -setMaxDepthInd 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72a5d27d-0a16-4e45-8282-bcfb26bac725",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --job-name=pcangsd_nomaf\n",
    "#SBATCH --time=03:00:00\n",
    "#SBATCH --mem=72GB\n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --output=%x_%A.out\n",
    "#SBATCH --error=%x_%A.err\n",
    "\n",
    "export PATH=$PATH:/$workdir/Software/pcangsd/\n",
    "\n",
    "INPUT=\"/$workdir/results/06_angsd/pcangsd/nomaf.beagle.gz\"\n",
    "OUTPUT=\"/$workdir/results/06_angsd/pcangsd/nomaf_pca\"\n",
    "\n",
    "pcangsd --beagle $INPUT --eig 2 --threads 20 --out $OUTPUT --selection --admix\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e07747-7240-4df0-b960-9787cce8afe8",
   "metadata": {},
   "source": [
    "# NGSadmix\n",
    "Also run with 1% and 5% MAF filters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df129950-fd6c-49cf-9314-84fa610ac976",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH --job-name=ngsadmix_nomaf\n",
    "#SBATCH --mem=70G\n",
    "#SBATCH --time=72:00:00\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --output=%x_%j.out\n",
    "#SBATCH --error=%x_%j.err\n",
    "#SBATCH --hint=nomultithread\n",
    "\n",
    "module purge\n",
    "module load angsd/0.935-GCC-9.2.0\n",
    "\n",
    "# Define variables\n",
    "BEAGLE_FILE=\"/$workdir/results/06_angsd/pcangsd/nomaf.beagle.gz\"\n",
    "OUT_PREFIX=\"/$workdir/results/06_angsd/pcangsd/ngsadmix_results/nomaf\"\n",
    "THREADS=${SLURM_CPUS_PER_TASK}\n",
    "MIN_MAF=0.00                 # enforced here to override default 5%\n",
    "MIS_TOL=0.8                  # tolerance for high-quality genotypes\n",
    "SEEDS=(14041 82680 21349)\n",
    "MAX_ITER=2000\n",
    "NGSadmix=\"/$workdir/Software/NGSadmix\"\n",
    "\n",
    "for K in {1..6}; do\n",
    "    echo \"Running ngsAdmix for K=${K}\"\n",
    "\n",
    "    # Loop through seeds\n",
    "    for SEED in \"${SEEDS[@]}\"; do\n",
    "        echo \"Using seed=${SEED}\"\n",
    "\n",
    "        # Run ngsAdmix\n",
    "        $NGSadmix -likes $BEAGLE_FILE \\\n",
    "                 -K $K \\\n",
    "                 -outfiles \"${OUT_PREFIX}_K${K}_seed${SEED}\" \\\n",
    "                 -seed $SEED \\\n",
    "                 -minMaf $MIN_MAF \\\n",
    "                 -misTol $MIS_TOL \\\n",
    "                 -P $THREADS \\\n",
    "                 -maxiter $MAX_ITER\n",
    "    done\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 (gimkl-2022a)",
   "language": "python",
   "name": "python3105-gimkl-2022a"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
