{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d53d5003-c76a-4bf3-a458-6de613cd8acc",
   "metadata": {},
   "source": [
    "# Sequencing processing and hardcalling variants with GATK"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99a98b3c-6516-40cb-918f-f7654c26ee82",
   "metadata": {},
   "source": [
    "## 1. Nextflow scripts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f23a9c0b-5087-4509-844e-40fa87250e9c",
   "metadata": {},
   "source": [
    "`Nextflow driver script`"
   ]
  },
  {
   "cell_type": "raw",
   "id": "9573a351-f8c7-4a7c-869a-0636fe05d4b6",
   "metadata": {},
   "source": [
    "#!/bin/bash -l\n",
    "#SBATCH --job-name=nf_pipeline\n",
    "#SBATCH --output=nf_pipeline_%j.out\n",
    "#SBATCH --error=nf_pipeline_%j.err\n",
    "#SBATCH --time=7-00:00:00\n",
    "#SBATCH --mem=1G\n",
    "#SBATCH --cpus-per-task=1\n",
    "\n",
    "# Create and move to working directory for job\n",
    "\n",
    "# load nextflow\n",
    "module load Nextflow/24.04.4\n",
    "\n",
    "# Copy files over to working directory\n",
    "BASE_DIR=/$Workdir/RRATTUS_ALL_SI\n",
    "\n",
    "# ensure gatk and picard know where to write temp files\n",
    "export _JAVA_OPTIONS=-Djava.io.tmpdir=/$Workdir/RRATTUS_ALL_SI/nf_temp\n",
    "\n",
    "nextflow -log ${BASE_DIR}/00_nextflow_pipeline_${SLURM_JOB_ID}.log \\\n",
    "run 00_variant_calling_pipeline.nf -resume \\\n",
    "-with-report $BASE_DIR/00_nextflow_pipeline_${SLURM_JOB_ID}_report.html -with-dag $BASE_DIR/00_nextflow_pipeline_${SLURM_JOB_ID}_flowchart.png"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e1d4414-09b8-468d-9dd4-f0a77ee5a3f4",
   "metadata": {},
   "source": [
    "`Nextflow script`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "754a6ff2-2a2b-4382-8562-3bdd3b91352a",
   "metadata": {},
   "outputs": [],
   "source": [
    "nextflow.enable.dsl = 2\n",
    "\n",
    "/*\n",
    " * pipeline input parameters\n",
    " */\n",
    "\n",
    "params.outdir = \"/$Workdir/RRATTUS_ALL_SI/results\"\n",
    "params.genome_index = \"/$Workdir/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic_autosomesonly.fna.gz\"\n",
    "\n",
    "process FASTQC {\n",
    "    module 'FastQC/0.12.1'\n",
    "    cpus 6\n",
    "    time '12h'\n",
    "    memory '16 GB'\n",
    "    queue 'large'\t\n",
    "    tag \"FASTQC on $sample_id\"\n",
    "    publishDir \"$params.outdir/\", mode:'copy'\n",
    "\n",
    "    input:\n",
    "    tuple val(sample_id), path(fastq_1), path(fastq_2)\n",
    "\n",
    "    output:\n",
    "    path \"01_fastqc/${sample_id}\"\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    mkdir -p 01_fastqc\n",
    "    mkdir -p 01_fastqc/${sample_id}\n",
    "    fastqc -o 01_fastqc/${sample_id} -f fastq -q ${fastq_1} ${fastq_2}\n",
    "    \"\"\"\n",
    "}\n",
    "\n",
    "process TRIMMING {\n",
    "    module 'TrimGalore/0.6.10-gimkl-2022a-Python-3.11.3-Perl-5.34.1'\n",
    "    cpus 16\n",
    "    time '24h'\n",
    "    memory '20 GB'\n",
    "    queue 'milan'\n",
    "    tag \"TRIMMING on $sample_id\"\n",
    "    publishDir \"$params.outdir/\", mode:'copy'\n",
    "\n",
    "    input:\n",
    "    tuple val(sample_id), path(fastq_1), path(fastq_2)\n",
    "\n",
    "    output:\n",
    "    path \"02_trimming/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    mkdir -p  02_trimming\n",
    "    mkdir -p 02_trimming/${sample_id}\n",
    "    trim_galore -q 20 -j 16 --fastqc -o 02_trimming/${sample_id}/ --paired \\\n",
    "    ${fastq_1} ${fastq_2}\n",
    "    \"\"\"\n",
    "}\n",
    "\n",
    "process MAPPING {\n",
    "    module 'BWA/0.7.17-GCC-11.3.0'\n",
    "    module 'SAMtools/1.16.1-GCC-11.3.0'\n",
    "    cpus 32\n",
    "    time '72h'\n",
    "    memory '75 GB'\n",
    "    queue 'milan'\n",
    "    tag \"MAPPING on $sample_id\"\n",
    "\n",
    "    input:\n",
    "    path \"02_trimming/${sample_id}\"\n",
    "    val(sample_id)\n",
    "   \n",
    "    output:\n",
    "    path \"03_mapped/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    mkdir -p 03_mapped\n",
    "    mkdir -p 03_mapped/${sample_id}\n",
    "    bwa mem -M -t 16 -R \"@RG\\\\tID:${sample_id}\\\\tSM:${sample_id}\\\\tPL:ILM\\\\tLB:${sample_id}\" \\\n",
    "    \"$params.genome_index\" \"02_trimming/${sample_id}/${sample_id}_R1_val_1.fq.gz\" \\\n",
    "    \"02_trimming/${sample_id}/${sample_id}_R2_val_2.fq.gz\" > \\\n",
    "    03_mapped/${sample_id}/${sample_id}.sam\n",
    "    samtools view -F 4 -Sb 03_mapped/${sample_id}/${sample_id}.sam | \n",
    "    samtools sort -o 03_mapped/${sample_id}/${sample_id}.bam\n",
    "    samtools index -c 03_mapped/${sample_id}/${sample_id}.bam\n",
    "    \"\"\"\n",
    "    \n",
    "}\n",
    "\n",
    "process DEDUPLICATION {\n",
    "    module 'GATK/4.4.0.0-gimkl-2022a'\n",
    "    module 'SAMtools/1.16.1-GCC-11.3.0'\n",
    "    cpus 24\n",
    "    time '24h'\n",
    "    memory '120 GB'\n",
    "    queue 'milan'\n",
    "    tag \"DEDUPLICATION on $sample_id\"\n",
    "    publishDir \"$params.outdir/\", mode:'copy'\n",
    "\n",
    "    input:\n",
    "    path \"03_mapped/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    output:\n",
    "    path \"05_deduplicated/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    mkdir -p 05_deduplicated\n",
    "    mkdir -p 05_deduplicated/${sample_id}\n",
    "    gatk --java-options -Xmx96g MarkDuplicates \\\n",
    "    -I \"03_mapped/${sample_id}/${sample_id}.bam\" \\\n",
    "    -O \"05_deduplicated/${sample_id}/${sample_id}_dedup.bam\" \\\n",
    "    -M \"05_deduplicated/${sample_id}/${sample_id}_metrics.txt\" \\\n",
    "    --MAX_RECORDS_IN_RAM 5000 -MAX_SEQS 5000 \\\n",
    "    --OPTICAL_DUPLICATE_PIXEL_DISTANCE 2500 \\\n",
    "    --VALIDATION_STRINGENCY SILENT \\\n",
    "    -MAX_FILE_HANDLES 1000\n",
    "    samtools index -c \"05_deduplicated/${sample_id}/${sample_id}_dedup.bam\"\n",
    "    \"\"\"\n",
    "\n",
    "}\n",
    "\n",
    "process GROUPING {\n",
    "    module 'SAMtools/1.16.1-GCC-11.3.0'\n",
    "    cpus 12\n",
    "    time '24h'\n",
    "    memory '30 GB'\n",
    "    queue 'milan'\n",
    "    tag \"GROUPING on $sample_id\"\n",
    "\n",
    "    input:\n",
    "    path \"05_deduplicated/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    output:\n",
    "    path \"06_grouped/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    mkdir -p 06_grouped\n",
    "    mkdir -p 06_grouped/${sample_id}\n",
    "    samtools addreplacerg -w -r 'ID:${sample_id}' -r 'LB:${sample_id}' -r 'SM:${sample_id}' -o \"06_grouped/${sample_id}/${sample_id}_grouped.bam\" \"05_deduplicated/${sample_id}/${sample_id}_dedup.bam\"\n",
    "    samtools index -c \"06_grouped/${sample_id}/${sample_id}_grouped.bam\"\n",
    "    \"\"\"\n",
    "\n",
    "}\n",
    "\n",
    "process HAPLOTYPECALLER {\n",
    "    module 'GATK/4.4.0.0-gimkl-2022a'\n",
    "    cpus 24\n",
    "    time '7d'\n",
    "    memory '100 GB'\n",
    "    queue 'milan'\n",
    "    tag \"HAPLOTYPECALLER on $sample_id\"\n",
    "    publishDir \"$params.outdir/\", mode:'copy'\n",
    "\n",
    "    input:\n",
    "    path \"06_grouped/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    output:\n",
    "    path \"07_haplotypecaller2/${sample_id}\"\n",
    "    val(sample_id)\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    mkdir -p 07_haplotypecaller2\n",
    "    mkdir -p 07_haplotypecaller2/${sample_id}\n",
    "    gatk --java-options \"-Xmx90g\" HaplotypeCaller \\\n",
    "   -R \"$params.genome_index\" \\\n",
    "   -I \"06_grouped/${sample_id}/${sample_id}_grouped.bam\" \\\n",
    "   -O \"07_haplotypecaller2/${sample_id}/${sample_id}_grouped.vcf.gz\" \\\n",
    "   -ERC GVCF \\\n",
    "   --create-output-variant-index true \\\n",
    "   --read-index \"06_grouped/${sample_id}/${sample_id}_grouped.bam.csi\"\n",
    "    \"\"\"\n",
    "\n",
    "}\n",
    "\n",
    "workflow {\n",
    "    Channel\n",
    "        .fromPath(\"/$Workdir/RRATTUS_ALL_SI/sample_file_list3.csv\")\n",
    "        .splitCsv(header: true)\n",
    "        .map {row -> tuple(row.sample_id,row.fastq_1,row.fastq_2)}\n",
    "        .set { sample_run_ch }\n",
    "\n",
    "    fastqc_ch = FASTQC( sample_run_ch )\n",
    "    trimming_ch = TRIMMING( sample_run_ch )\n",
    "    mapping_ch = MAPPING( trimming_ch )\n",
    "    deduplication_ch = DEDUPLICATION( mapping_ch )\n",
    "    grouping_ch = GROUPING( deduplication_ch )\n",
    "    haplotypecaller_ch = HAPLOTYPECALLER( grouping_ch )\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2be3918-363a-4742-a1a2-6aba058e387a",
   "metadata": {},
   "source": [
    "## 2. Combine GVCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9bd1a5b-a3d3-4e0d-88ce-07783f26de0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --job-name=combineGVCF\n",
    "#SBATCH --time=4-00:00:00\n",
    "#SBATCH --mem=200GB\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH --output=%x_%A.out\n",
    "#SBATCH --error=%x_%A.err\n",
    "\n",
    "#load mods\n",
    "module load GATK\n",
    "\n",
    "# set args\n",
    "REFERENCE_GENOME=\"/$Workdir/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic_autosomesonly.fna.gz\"\n",
    "WD=\"/$Workdir/RRATTUS_ALL_SI/results/08_combine\"\n",
    "GVCFS=\"/$Workdir/RRATTUS_ALL_SI/results/07_haplotypecaller2\"\n",
    "\n",
    "gvcf_list=\"\"\n",
    "for vcf_file in ${GVCFS}/{01..79}/*.vcf.gz; do\n",
    "    echo \"Processing: ${vcf_file}\"\n",
    "    gvcf_list+=\" -V ${vcf_file}\"\n",
    "done\n",
    "\n",
    "echo \"GVCF List: ${gvcf_list}\"\n",
    "\n",
    "gatk --java-options \"-Xmx180g\" CombineGVCFs \\\n",
    "    -R $REFERENCE_GENOME \\\n",
    "    ${gvcf_list} \\\n",
    "    -O \"${WD}/multisample.g.vcf.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d270d91d-2c68-4cd3-a8da-e5810bdb10ee",
   "metadata": {},
   "source": [
    "## 3. Genotype GVCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ebf308-cd04-4236-9748-75d64987fab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --job-name=genotypeGVCF\n",
    "#SBATCH --time=5-00:00:00\n",
    "#SBATCH --mem=200GB\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH --array=1-27\n",
    "#SBATCH --output=genotypeGVCF_%A_%a.out\n",
    "#SBATCH --error=genotypeGVCF_%A_%a.err\n",
    "\n",
    "#load mods\n",
    "module load GATK\n",
    "\n",
    "# set args\n",
    "REFERENCE_GENOME=\"/$Workdir/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/GCF_011064425.1_Rrattus_CSIRO_v1_genomic_autosomesonly.fna.gz\"\n",
    "OUTPUT=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/multisample_allvariants_gatk\"\n",
    "INPUT=\"/$Workdir/RRATTUS_ALL_SI/results/08_combine/multisample.g.vcf.gz\"\n",
    "INTERVAL_FILE=\"/$Workdir/reference_sequences/GCF_011064425.1_Rrattus_CSIRO_v1/interval_scattergenotyping.list\"\n",
    "\n",
    "# match a line number with array job\n",
    "INTERVAL=$(sed -n \"${SLURM_ARRAY_TASK_ID}p\" \"$INTERVAL_FILE\")\n",
    "\n",
    "echo \"Processing: $INTERVAL\"\n",
    "\n",
    "gatk --java-options \"-Xmx180g -XX:ParallelGCThreads=64\" GenotypeGVCFs \\\n",
    "    -R $REFERENCE_GENOME \\\n",
    "    -L $INTERVAL \\\n",
    "    -V $INPUT \\\n",
    "    -O \"${OUTPUT}_${INTERVAL}.vcf.gz\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7cd67fd-6f12-4bb3-88c4-add1c9954f7d",
   "metadata": {},
   "source": [
    "## 4. Gather variants by sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8af0cd88-ef85-4682-9beb-c37b02f6a0d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --job-name=concatbcftools_allowovers\n",
    "#SBATCH --time=06:00:00\n",
    "#SBATCH --mem=32GB\n",
    "#SBATCH --cpus-per-task=2\n",
    "#SBATCH --output=%x_%A.out\n",
    "#SBATCH --error=%x_%A.err\n",
    "\n",
    "# load mods\n",
    "module load BCFtools\n",
    "\n",
    "VCF_FILELIST=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/vcf_filelist.txt\"\n",
    "OUTPUT=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/multisample_allvars_gatk_combined.vcf.gz\"\n",
    "\n",
    "bcftools concat -f $VCF_FILELIST --allow-overlaps -Oz -o $OUTPUT"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c9e944a-be36-4d69-8d1d-9d336cd4f709",
   "metadata": {},
   "source": [
    "## 5. Filtering (BCFtools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204a274b-379a-40dd-9c05-5407b8c6c064",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH --account=$user\n",
    "#SBATCH --job-name=bcftoolsfiltersnps\n",
    "#SBATCH --time=06:00:00\n",
    "#SBATCH --mem=24GB\n",
    "#SBATCH --cpus-per-task=4\n",
    "#SBATCH --output=%x_%A.out\n",
    "#SBATCH --error=%x_%A.err\n",
    "\n",
    "# load mods\n",
    "module purge\n",
    "module load GATK\n",
    "module load BCFtools/1.16-GCC-11.3.0\n",
    "\n",
    "INPUT=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/multisample_allvars_gatk_combined.vcf.gz\"\n",
    "OUTPUT_SNP_VCF=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/multisample_allvariants_gatk_combined_SNPs.vcf.gz\"\n",
    "OUTPUT_SNP_VCF_FILT=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/multisample_allvariants_gatk_combined_SNPsfiltered.vcf.gz\"\n",
    "OUTPUT_FINAL=\"/$Workdir/RRATTUS_ALL_SI/results/09_genotyped_GATK/multisample_allvariants_gatk_combined_SNPsfiltered_final.vcf.gz\"\n",
    "\n",
    "gatk --java-options \"-Xmx22g -XX:ParallelGCThreads=8\" IndexFeatureFile -I $INPUT\n",
    "\n",
    "echo \"gatk indexing complete.\"\n",
    "\n",
    "# Select biallelic SNPs\n",
    "gatk --java-options \"-Xmx22g -XX:ParallelGCThreads=8\" SelectVariants \\\n",
    "   -V $INPUT \\\n",
    "   -select-type SNP \\\n",
    "   -restrict-alleles-to BIALLELIC \\\n",
    "   -O $OUTPUT_SNP_VCF\n",
    "\n",
    "echo \"Biallelic SNP selection completed. Output file: $OUTPUT_SNP_VCF\"\n",
    "\n",
    "# min sample read depth = 4 and max = 100\n",
    "# set failed genotypes to missing\n",
    "bcftools filter -i 'FMT/DP>4 & FMT/DP<100 & FMT/GQ>20' -S . -Oz -o $OUTPUT_SNP_VCF_FILT $OUTPUT_SNP_VCF\n",
    "\n",
    "# keep only the snps that have passed out GATK filters (i.e., FILTER=PASS)\n",
    "bcftools view -f PASS $OUTPUT_SNP_VCF_FILT > $OUTPUT_FINAL\n",
    "\n",
    "echo \"Keeping only variants with filter=PASS. Output file: $OUTPUT_FINAL\"\n",
    "\n",
    "rm $OUTPUT_SNP_VCF\n",
    "rm $OUTPUT_SNP_VCF_FILT\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.5 (gimkl-2022a)",
   "language": "python",
   "name": "python3105-gimkl-2022a"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
